sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,290 @@
+	bucketRoot := s.BucketStorageView.Prefix()
+
+	if strings.HasPrefix(path, bucketRoot) {
+		return strings.TrimPrefix(path, bucketRoot), nil
+	} else {
+		return "", fmt.Errorf("path %q doesn't match bucket storage %q", path, bucketRoot)
+	}
+}
+
+// InvalidateItems handles invalidation of a StoragePackerV2 bucket during replication.
+//
+// Given a storage path (to a bucket) and new contents, return
+// "present": all items present in the new bucket (new, changed, or unchanged)
+// "deleted": all items that were in the cache but are now absent
+//
+// For normal WAL operation calls to this function are serialized based on the order PutItem
+// wrote the buckets; for recovery from the Merkle tree this will not be the case.
+//
+// For each bucket B we receive, process the items in it, compared with items in
+// the previous version of B in the cache (if any, call it C) and elsewhere in the current cache.
+//
+// If an item exists in B, report it in "present" even if it also exists elsewere in the cache
+//
+// If an item exists in C, but not in B, then it might have been deleted.
+//    * If the item exists elsewhere in the cache (might be a child, might even be a parent)
+//      don't mark it as deleted. Either we will do it later (when we process that bucket)
+//      or the item is now in its correct location.
+//         * One problem here is that we can't rely upon the longest suffix match to find the bucket.
+//         * Another problem is that the local side might have a newer (or older) version in its
+//           memory compared to storage. If we can't be confident that InvalidateItems() will be
+//           later called on the bucket that *does* have the consistent version, we have to report any
+//           such item as "present" to give the upper layer a chance to look at it.
+//    * If the item does not exist elsewhere in the cache, report it as deleted.
+//
+// If an item has been moved by sharding, it's possible this will cause it to be reported
+// as deleted and re-created.
+//
+// Either B or C might be nonexistent; just treat those as empty.
+//
+// If we're the secondary, we shouldn't have to worry about concurrent access.
+func (s *StoragePackerV2) InvalidateItems(ctx context.Context, path string, newValue []byte) (present []*Item, deleted []*Item, err error) {
+	// Double-check that it's on our path, and figure out which bucket it should be
+	bucketKey, err := s.bucketKeyFromPath(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var replacementBucket *LockedBucket
+	bucketRemoved := false
+
+	if newValue == nil {
+		bucketRemoved = true
+		// Create a placeholder so we can run the rest of the logic
+		// against an empty collection
+		replacementBucket = s.newEmptyBucket(bucketKey)
+	} else {
+		// Fake up a storage entry for the bucket and decode it
+		storage := &logical.StorageEntry{
+			Value: newValue,
+		}
+		replacementBucket, err = s.DecodeBucket(storage)
```

Comment:
Wouldn't the cacheKey in the Decode call result in an empty `cacheKey`? I think it'd lead to weird locking behavior due to this value being empty in `locksutil.LockForKey(s.storageLocks, cacheKey)`.

Reply:
I believe you're correct, good catch.

DecodeBucket should probably use the demarshaled bucket key rather than the storage key.

Reply:
OK, fixed 2x (and made sure both pass the new test.)

1. DecodeBucket uses the Bucket's key if it's present
2. InvalidateItems provides a valid Key in the StorageEntry